### PR TITLE
HIVE-28337: Process timestamps at UTC timezone instead of local timezone in MetaStoreUtils

### PR DESCRIPTION
**Pull request by KiranVelumuri**
### What changes were proposed in this pull request?

Timestamps are processed using LocalDateTime at local time zone. Change this to use Instant instead to process timestamps at UTC time zone.

### Why are the changes needed?

https://issues.apache.org/jira/browse/HIVE-28337

### Does this PR introduce _any_ user-facing change?

No

### Is the change a dependency upgrade?

No

### How was this patch tested?

mvn test -Dtest=TestMetaStoreUtils

Before the patch:
with Random(seed: 23) in line 65(default case)
image

with Random(seed: 24) in line 65
image

After the patch:
image

### CODE REVIEW

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java
Comment (wecharyu) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java:56]: How about keeping `LocalDateTime` in constructor parameter, and generate the timeZone related timestamp string here:
```java
this.timestamp = ZonedDateTime.of(timestamp, ZoneId.of(zoneId)).format(FORMATTER);
```

**Comment by wecharyu** [comment]
Nice catch~ This failed test was also found in https://github.com/apache/hive/pull/5004 and we can revert https://github.com/kgyrtkirk/hive-dev-box/pull/16 once this patch is merged. 
@zabetak could you also take a look?

**Review by zabetak — CHANGES_REQUESTED** [review]
The intention of the test is to ensure that conversions from/to string do not alter the value no matter the timezone because the DATE/TIMESTAMP data types [should be timezone agnostic|https://cwiki.apache.org/confluence/display/Hive/Different+TIMESTAMP+types].

Changing LocalDateTime to ZonedDateTime defeats the purpose of the tests and actually hides a bug that affects the MetaStoreUtils APIs.

I left a similar comment under the JIRA ticket.

**Comment by KiranVelumuri** [comment]
The 2 test cases mentioned in https://issues.apache.org/jira/browse/HIVE-28337 are failing for me locally in my machine with Java version 1.8.0_331. I tried running these test cases in https://github.com/apache/hive/pull/5004, but the DST related one was successful. I was investigating as to why this was happening, and now I get it seeing the previous comments that this is occurring due to JDK bug.

**Comment by KiranVelumuri** [comment]
> The intention of the test is to ensure that conversions from/to string do not alter the value no matter the timezone because the DATE/TIMESTAMP data types [should be timezone agnostic](https://cwiki.apache.org/confluence/display/Hive/Different+TIMESTAMP+types).
> 
> Changing LocalDateTime to ZonedDateTime defeats the purpose of the tests and actually hides a bug that affects the MetaStoreUtils APIs.
> 
> I left a similar comment under the JIRA ticket.

Yes, I agree with you regarding the intention and purpose of the test. I am looking into the issue and would re-work this PR taking the MetaStoreUtils APIs into account.

**Comment by KiranVelumuri** [comment]
I have updated the tests and the MetaStoreUtils APIs to always have Timestamp created in UTC timezone, since the timestamps during the daylight savings transition could not be represented in the local time zone.


However, there are 2 known issues for the timestamps mentioned below but it is the expected behaviour from Java's Timestamp:
1. Timestamps with year 0000 would be converted to year 0001.
2. Timestamps in the range [1582-10-05 00:00:00, 1582-10-14 23:59:59] (both inclusive) would have 10 days added due to the shift from Julian calendar to Gregorian calendar. (The next day after 1582-10-04 is 1582-10-15)

References:
[1] https://www.ibm.com/docs/en/db2/11.5?topic=dttmddtija-date-time-timestamp-values-that-can-cause-problems-in-jdbc-sqlj-applications#imjcc_r0053436__title__5
[2] https://strasbourg.eleusal.com/en/what-happened-between-5-and-14-october-1582/

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java
Comment (KiranVelumuri) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java]: From @zabetak's comment, I have reverted to LocalDateTime.

**Review by zabetak — CHANGES_REQUESTED** [review]
I understand that due to bugs the tests may fail in certain environments.

If the main motivation of this PR is to avoid the failures when this happens it seems more appropriate to add conditions for detecting skipping the specific test with appropriate bug reference. 

If the intention is to fix the problems revealed by the tests then more work is needed since the proposed changes cannot land as they are.

**Comment by KiranVelumuri** [comment]
g

**Comment by KiranVelumuri** [comment]
I have updated the test to not hit the failing cases. Will raise a separate Jira for the rework that is needed around timestamp. @wecharyu @zabetak Please review. Thank you.

**Comment by KiranVelumuri** [comment]
> I understand that due to bugs the tests may fail in certain environments.
> 
> If the main motivation of this PR is to avoid the failures when this happens it seems more appropriate to add conditions for detecting skipping the specific test with appropriate bug reference.
> 
> If the intention is to fix the problems revealed by the tests then more work is needed since the proposed changes cannot land as they are.

The initial motivation was to fix the test failures, which later changed to solve the problem revealed in MetaStoreUtils. Upon your comments, I have reworked my PR which now solves the problem in MetaStoreUtils and accordingly modified TestMetaStoreUtils to validate the tests properly.

@zabetak @wecharyu Please review. Thank you.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreUtils.java
Comment (zabetak) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/utils/MetaStoreUtils.java:137]: Since now we are converting to `Instant` do we need to pass from `LocalDateTime`? Can we use directly `Instant.from` here?

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java
Comment (zabetak) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java:85]: These 5 tests seem kinda identical maybe we just need one of them.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java
Comment (zabetak) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java:81]: Instead of using the `LocalDateTime.ofEpochSecond` constructor you could use the basic constructor which allows you to specify all fields explicitly. By doing this the comment above becomes redundant since it is clear from the code what is the date.

**Review by zabetak — COMMENTED** [review]
LGTM. I left some small comments but hopefully nothing major. Let's see how the tests go.

@KiranVelumuri Please update the PR/JIRA summary and description to focus on the changes that were done in the production (and why) rather than the test failures.

@mdayakar @dengzhhu653 This is related to code that you touched recently so you may want to take a look.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java
Comment (zabetak) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java]: This comment is not relevant in the latest commit. Please ignore.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java
Comment (KiranVelumuri) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/TestMetaStoreUtils.java]: They are in different time zones with DST, so included them

**Review by zabetak — APPROVED** [review]
LGTM, will merge soon.

**Pull request closed without merging**
